fe/fe-core/src/main/java/org/apache/doris/nereids/util/ExpressionUtils.java

```diff
@@ -89,51 +87,36 @@ public static Expression or(List<Expression> expressions) {
      * Use AND/OR to combine expressions together.
      */
     public static Expression combine(NodeType op, List<Expression> expressions) {
-
         Objects.requireNonNull(expressions, "expressions is null");
 
         if (expressions.size() == 0) {
-            if (op == NodeType.AND) {
-                return new Literal(true);
-            }
-            if (op == NodeType.OR) {
-                return new Literal(false);
-            }
-        }
-
-        if (expressions.size() == 1) {
+            return new Literal(op == NodeType.AND);
+        } else if (expressions.size() == 1) {
             return expressions.get(0);
         }
 
-        List<Expression> distinctExpressions = Lists.newArrayList(new LinkedHashSet<>(expressions));
-        if (op == NodeType.AND) {
-            if (distinctExpressions.contains(Literal.FALSE_LITERAL)) {
-                return Literal.FALSE_LITERAL;
+        Expression shortCircuit = (op == NodeType.AND ? Literal.FALSE_LITERAL : Literal.TRUE_LITERAL);
+        Expression skip = (op == NodeType.AND ? Literal.TRUE_LITERAL : Literal.FALSE_LITERAL);
+        LinkedHashSet<Expression> distinctExpressions = Sets.newLinkedHashSetWithExpectedSize(expressions.size());
+        for (Expression expression : expressions) {
+            if (expression.equals(shortCircuit)) {
+                return shortCircuit;
+            } else if (!expression.equals(skip)) {
+                distinctExpressions.add(expression);
             }
-            distinctExpressions = distinctExpressions.stream().filter(p -> !p.equals(Literal.TRUE_LITERAL))
-                    .collect(Collectors.toList());
         }
 
-        if (op == NodeType.OR) {
-            if (distinctExpressions.contains(Literal.TRUE_LITERAL)) {
-                return Literal.TRUE_LITERAL;
+        List<Expression> result = Lists.newArrayListWithCapacity(distinctExpressions.size() / 2 + 1);
```

Comment:
maybe we could use stream reduce api to do this without recursion

Reply:
The output would not be the same as the original one if we used `stream reduce` api.

recursion: `(A AND B) AND (C AND D)`
reduce: `((A AND B) AND C) AND D)`

If the inconsistency is acceptable, I will use the `reduce` way to simplify it.

Reply:
Simplified it by `reduce` api.